File: dsl/src/main/java/cbs/dsl/builder/EventBuilder.java
package cbs.dsl.builder;

import cbs.dsl.api.EventDefinition;
import cbs.dsl.api.context.DisplayScope;
import cbs.dsl.api.context.EnrichmentContext;
import cbs.dsl.api.context.FinishContext;
import cbs.dsl.api.context.TransactionsScope;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.BiConsumer;
import java.util.function.Consumer;

/**
 * Builder for creating {@link EventDefinition} instances.
 *
 * <p>This is a convenience DSL — it describes wiring but does not execute anything.
 */
public class EventBuilder {

  private final String code;
  private final List<String> transactionCodes = new ArrayList<>();
  private Consumer<EnrichmentContext> contextBlock = ctx -> {};
  private Consumer<DisplayScope> displayBlock = scope -> {};
  private BiConsumer<FinishContext, Throwable> finishBlock = (ctx, ex) -> {};

  EventBuilder(String code) {
    this.code = code;
  }

  /**
   * Adds a transaction code to the event's execution sequence.
   *
   * @param transactionCode the transaction code to add
   * @return this builder for chaining
   */
  public EventBuilder transaction(String transactionCode) {
    this.transactionCodes.add(transactionCode);
    return this;
  }

  /**
   * Sets the context enrichment block.
   *
   * @param block the context block
   * @return this builder for chaining
   */
  public EventBuilder context(Consumer<EnrichmentContext> block) {
    this.contextBlock = block;
    return this;
  }

  /**
   * Sets the display block.
   *
   * @param block the display block
   * @return this builder for chaining
   */
  public EventBuilder display(Consumer<DisplayScope> block) {
    this.displayBlock = block;
    return this;
  }

  /**
   * Sets the finish block.
   *
   * @param block the finish block
   * @return this builder for chaining
   */
  public EventBuilder finish(BiConsumer<FinishContext, Throwable> block) {
    this.finishBlock = block;
    return this;
  }

  /**
   * Builds an immutable {@link EventDefinition}.
   *
   * @return the built event definition
   */
  public EventDefinition build() {
    List<String> txCodes = Collections.unmodifiableList(new ArrayList<>(transactionCodes));
    return new EventDefinition() {
      @Override
      public String getCode() {
        return code;
      }

      @Override
      public Consumer<EnrichmentContext> getContextBlock() {
        return contextBlock;
      }

      @Override
      public Consumer<DisplayScope> getDisplayBlock() {
        return displayBlock;
      }

      @Override
      public Consumer<TransactionsScope> getTransactionsBlock() {
        if (txCodes.isEmpty()) {
          return null;
        }
        return scope -> {
          for (String txCode : txCodes) {
            // In a generated workflow, this would be replaced by hardcoded activity calls.
            // The builder scope is for local/dev execution only.
          }
        };
      }

      @Override
      public BiConsumer<FinishContext, Throwable> getFinishBlock() {
        return finishBlock;
      }
    };
  }
}

File: dsl/src/main/kotlin/cbs/dsl/runtime/EventBuilder.kt
package cbs.dsl.runtime

import cbs.dsl.api.EventDefinition
import cbs.dsl.api.ParameterDefinition
import cbs.dsl.api.context.DisplayScope
import cbs.dsl.api.context.EnrichmentContext
import cbs.dsl.api.context.FinishContext
import cbs.dsl.api.context.TransactionsScope

class EventBuilder(
    override val code: String,
) : EventDefinition {
    private val _parameters = mutableListOf<ParameterDefinition>()
    override val parameters: List<ParameterDefinition> get() = _parameters.toList()

    private var _contextBlock: (EnrichmentContext) -> Unit = {}
    override val contextBlock: (EnrichmentContext) -> Unit get() = _contextBlock

    private var _displayBlock: (DisplayScope) -> Unit = {}
    override val displayBlock: (DisplayScope) -> Unit get() = _displayBlock

    private var _transactionsBlock: (suspend TransactionsScope.() -> Unit)? = null
    override val transactionsBlock: (suspend TransactionsScope.() -> Unit)? get() = _transactionsBlock

    private var _finishBlock: (FinishContext, Throwable?) -> Unit = { _, _ -> }
    override val finishBlock: (FinishContext, Throwable?) -> Unit get() = _finishBlock

    fun parameters(block: ParametersScope.() -> Unit) {
        _parameters += ParametersScope().apply(block).definitions
    }

    fun context(block: (EnrichmentContext) -> Unit) {
        _contextBlock = block
    }

    fun display(block: (DisplayScope) -> Unit) {
        _displayBlock = block
    }

    fun transactions(block: suspend TransactionsScope.() -> Unit) {
        _transactionsBlock = block
    }

    fun finish(block: (FinishContext, Throwable?) -> Unit) {
        _finishBlock = block
    }
}

fun event(
    code: String,
    block: EventBuilder.() -> Unit,
): EventDefinition = EventBuilder(code).apply(block)
